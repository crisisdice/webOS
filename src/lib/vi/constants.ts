export enum VI_MODE {
  VISUAL,
  INSERT,
  COMMAND,
  VISUAL_BLOCK,
  VISUAL_LINE,
}
