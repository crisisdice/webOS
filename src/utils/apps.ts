export enum APPS {
  SH,
  VI,
}
