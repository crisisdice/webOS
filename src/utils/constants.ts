export const EMPTY = ''

export const ENV = 'ENV'

export const FS = 'FS'

export const PWD = 'PWD'

export const HOME = 'HOME'

export const ARROW_KEYS = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight']
