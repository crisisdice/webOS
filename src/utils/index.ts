export * from './dom'
export * from './state'
export * from './constants'
export * from './apps'
