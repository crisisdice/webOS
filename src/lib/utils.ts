import type { Line } from './types'
import { DASH } from './constants'

export const separate = (args: string): { name: string; path: string } => {
  if (args.indexOf('/') === -1) return { path: '.', name: args }
  const tokens = args.split('/')
  const path = tokens.slice(0, tokens.length - 1).join('/')
  return { name: tokens.at(-1), path }
}

export const refocus = (e: Event) =>
  window.setTimeout(() => (e.target as HTMLInputElement).focus(), 0)

export const lineToString = ({ PRECARET, CARET, POSTCARET }: Line) => {
  return `${PRECARET}${CARET === DASH ? '' : CARET}${POSTCARET}`
}
