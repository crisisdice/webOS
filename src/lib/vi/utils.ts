import type { Line, ViState } from '../../types'
import { write, stat } from '../fs'
import { separate } from '../path'
import { EMPTY_LINE, INITAL_STATE, bufferToFile, lineToString, EMPTY } from '../../utils'
import { VI_MODE } from './constants'

function parseCmd(input: Line) {
  const commandTokens = lineToString(input).split(' ')
  const cmd = commandTokens?.[0] ?? ''
  const args = commandTokens.slice(1)

  const cmds = cmd.split('')
  const override = cmds.includes('!')

  return { cmds, args, override }
}

function writeFile({ args, STATE }: { args: string[]; STATE: ViState }): ViState {
  let { FILE_NAME } = STATE
  const { BUFFER } = STATE
  FILE_NAME = FILE_NAME ?? args?.[0]

  if (!FILE_NAME) return report({ STATE, MESSAGE: 'No file name' })

  write({ ...separate(FILE_NAME), obj: bufferToFile(BUFFER) })
  return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL, FILE_NAME }
}

function checkSaveStatus({ STATE }: { STATE: ViState }): boolean {
  const { FILE_NAME, BUFFER } = STATE
  const { exists, obj } = stat(FILE_NAME)
  const file = bufferToFile(BUFFER)

  return !exists ? file === EMPTY : obj === file
}

function report({ STATE, MESSAGE }: { STATE: ViState; MESSAGE: string }) {
  return { ...STATE, MESSAGE, COMMAND_LINE: EMPTY_LINE }
}

export function vimCommands(STATE: ViState) {
  const { COMMAND_LINE, OLD_COMMANDS } = STATE
  const { cmds, args, override } = parseCmd(COMMAND_LINE)

  for (const cmd of cmds) {
    if (cmd === 'w') {
      STATE = writeFile({ args, STATE })
      continue
    }

    if (cmd === 'q') {
      if (!checkSaveStatus({ STATE }) && !override) {
        return report({ STATE, MESSAGE: 'No write since last change' })
      }
      return { ...INITAL_STATE, OLD_COMMANDS }
    }

    return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL }
  }
  return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL }
}
