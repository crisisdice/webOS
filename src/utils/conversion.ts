import type { Line, ViState } from '../types'
import { EMPTY, PWD, EMPTY_BUFFER } from './constants'
import { getEnv, stat } from '../lib'

export const lineToString = ({ PRECARET, CARET, POSTCARET }: Line) => {
  return `${PRECARET}${CARET === null ? '' : CARET}${POSTCARET}`
}

export const stringToLine = (line: string, x: number, CARET_WIDTH = 1): Line => {
  const PRECARET = line.slice(0, x)
  const EOL = x === line.length
  const CARET = EOL || CARET_WIDTH === 0 ? null : line[x]
  const POSTCARET = EOL ? EMPTY : line.slice(x + CARET_WIDTH)

  return {
    PRECARET,
    CARET,
    POSTCARET,
    EOL,
    CARET_WIDTH,
  }
}

export function fileToBuffer(fileName: string | null): {
  BUFFER: ViState['BUFFER']
  FILE_NAME: string | null
} {
  const { exists, obj, path, isDirectory } = stat('', fileName)

  if (!exists || isDirectory) {
    return { BUFFER: EMPTY_BUFFER, FILE_NAME: fileName ? `${getEnv(PWD)}/${fileName}` : null }
  }
  if (obj === EMPTY) {
    return { BUFFER: EMPTY_BUFFER, FILE_NAME: path }
  }

  const file = (obj as string).split('\n')

  const BUFFER_PRE = []
  const LINE = stringToLine(file[0], 0)
  const BUFFER_POST = file.slice(1)

  return { BUFFER: { BUFFER_PRE, BUFFER_POST, LINE }, FILE_NAME: path }
}
