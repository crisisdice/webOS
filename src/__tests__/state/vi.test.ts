import { run, fill } from '../helpers'
import { BACK, DOWN, ENTER, LEFT, RIGHT, UP } from '../constants'
import { init } from '../../utils'

const ENTER_VI = [...'vi'.split(''), ENTER]

const tests = [
  [...ENTER_VI, 'i', ...'hello'.split('')],
  [...ENTER_VI, 'i', ...'hello'.split(''), ...fill(6, LEFT), ...fill(6, RIGHT)],
  [...ENTER_VI, 'i', ...'hello'.split(''), '`'],
  [...ENTER_VI, 'i', ...'hello'.split(''), '`', ':', ...'cmd'],
  [...ENTER_VI, 'i', ...'hello'.split(''), '`', ':', ...'cmd', ...fill(6, LEFT), ...fill(6, RIGHT)],
  [...ENTER_VI, 'i', ...'hello'.split(''), ENTER, ...'hello world'.split(''), UP],
  [
    ...ENTER_VI,
    'i',
    ...'hello'.split(''),
    ENTER,
    ...'hello world'.split(''),
    UP,
    DOWN,
    ...fill(12, BACK),
  ],
]

beforeEach(() => init())

for (const sequence of tests) {
  run(sequence)
}
