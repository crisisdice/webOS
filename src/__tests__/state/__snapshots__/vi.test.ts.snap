// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 3> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [
      "",
      "",
    ],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 2,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 3> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 3> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 5> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 5> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER><BACK 5> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUp 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [
      "",
      "",
      "",
    ],
    "BUFFER_PRE": [
      "",
      "",
    ],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 2,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUp 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUp 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUpArrowUpArrowUp 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [
      "",
      "",
      "",
      "",
      "",
    ],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUpArrowUpArrowUp 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>i<ENTER><ENTER><ENTER><ENTER><ENTER>ArrowUpArrowUpArrowUpArrowUpArrowUp 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "hello",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 5,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world<LEFT 5><ENTER> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [
      "hello ",
    ],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": false,
      "POSTCARET": "world",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 1,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world<LEFT 5><ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello world<LEFT 5><ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world\`:q!<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world\`:q!<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello world\`:q!<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world\`:q<ENTER> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "d",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hello worl",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 10,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": "No write since last change",
  "MODE": 2,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world\`:q<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello world\`:q<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world\`:w out<ENTER>\`:q<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world\`:w out<ENTER>\`:q<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":"","out":"hello world"},"root":{}}}"`;

exports[`vi<ENTER>ihello world\`:w out<ENTER>\`:q<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world\`:w<ENTER> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "d",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hello worl",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 10,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": "No file name",
  "MODE": 0,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world\`:w<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello world\`:w<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello world\`:wq<ENTER> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "d",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hello worl",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 10,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": "No write since last change",
  "MODE": 2,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello world\`:wq<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello world\`:wq<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUp 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [
      "hello world",
    ],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "hello",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 5,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUp 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUp 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUpArrowDown<BACK 12> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": false,
      "POSTCARET": " world",
      "PRECARET": "",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 0,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUpArrowDown<BACK 12> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello<ENTER>hello worldArrowUpArrowDown<BACK 12> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello<LEFT 6><RIGHT 6> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": null,
      "CARET_WIDTH": 0,
      "EOL": true,
      "POSTCARET": "",
      "PRECARET": "hello",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 5,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 1,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello<LEFT 6><RIGHT 6> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello<LEFT 6><RIGHT 6> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello\` 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "o",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hell",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 4,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 0,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello\` 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello\` 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello\`:cmd 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "o",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hell",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "cmd",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 4,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 2,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello\`:cmd 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello\`:cmd 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`vi<ENTER>ihello\`:cmd<LEFT 6><RIGHT 6> 1`] = `
{
  "BUFFER": {
    "BUFFER_POST": [],
    "BUFFER_PRE": [],
    "LINE": {
      "CARET": "o",
      "CARET_WIDTH": 1,
      "EOL": false,
      "POSTCARET": "",
      "PRECARET": "hell",
    },
  },
  "COMMAND_LINE": {
    "CARET": null,
    "CARET_WIDTH": 1,
    "EOL": true,
    "POSTCARET": "",
    "PRECARET": "cmd",
  },
  "CONTROL_DOWN": false,
  "COORDS": {
    "x": 4,
    "y": 0,
  },
  "DOWN_MAPPING": [Function],
  "FILE_NAME": null,
  "HISTORY_INDEX": -1,
  "MESSAGE": null,
  "MODE": 2,
  "NAME": 1,
  "OLD_COMMANDS": [
    {
      "COMMAND_LINE": {
        "CARET": null,
        "CARET_WIDTH": 1,
        "EOL": true,
        "POSTCARET": "",
        "PRECARET": "vi",
      },
      "PWD": "/home/guest",
      "STDOUT": "",
      "USER": "guest",
    },
  ],
  "PWD": "/home/guest",
  "REGISTERS": {},
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`vi<ENTER>ihello\`:cmd<LEFT 6><RIGHT 6> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`vi<ENTER>ihello\`:cmd<LEFT 6><RIGHT 6> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;
